### PR TITLE
CORE-4868 - correct version handling for merging CPK entities

### PR DESCRIPTION
**Pull request by conal-smith**
before merging CPI and CPK entities we must find the entities first by their primary keys and if the entities exist, update these before merging rather than creating a new entity each time with the default entityVersion = 0

### CODE REVIEW

**Review by vkolomeyko — DISMISSED** [review]
LGTM

**Review by vkolomeyko — APPROVED** [review]
LGTM

File: components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt [hunk 1]
```diff
@@ -355,26 +348,87 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
...
+        existingCpiCpkEntity.metadata.update(
+            cpk.metadata.fileChecksum.toString(),
+            "1",
+            cpk.metadata.toJsonAvro(),
+            false
+        )
+    )
+
+    private fun createOrUpdateExistingCpkFileEntities(em: EntityManager, cpks: Collection<Cpk>): List<CpkFileEntity> {
+        val query = """
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:407]: I don't think we should be doing this because this will query the data inlcuding the binary which we don't need.
In order to get a JPA managed entity to update, I think we can use `getReference`.

File: components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt [hunk 2]
```diff
@@ -28,11 +28,15 @@ import javax.persistence.EntityManager
...
 import javax.persistence.NonUniqueResultException
+import net.corda.libs.cpi.datamodel.CpiCpkEntity
+import net.corda.libs.cpi.datamodel.CpiCpkKey
 import net.corda.libs.cpi.datamodel.CpkKey
+import net.corda.libs.packaging.core.CpkIdentifier
 
 /**
  * This class provides some simple APIs to interact with the database.
  */
+@Suppress("TooManyFunctions")
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:39]: Possibly a bit out of scope for this PR, but this class has become quite big and quite hard to follow.
Some of the comments were taken out, which is ok, but it hasn't helped the readability.
I think initially this was quite generic, but isn't anymore, maybe we need to break this down a bit.

File: components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt [hunk 3]
```diff
@@ -229,21 +233,16 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
         requestId: RequestId,
         groupId: String
     ): CpiMetadataEntity {
-        val cpiMetadataEntity = createCpiMetadataEntity(cpi, cpiFileName, checksum, requestId, groupId)
         entityManagerFactory.createEntityManager().transaction { em ->
-            // persist metadata
+            val cpkMetadataEntities = createOrUpdateCpkMetadataEntities(cpi, emptyMap())
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:237]: Is there still a need for 2 separate functions (persist and update)? Initially the function was called `persist` as the intention was that it would support both insert and update. I think there is sufficient overlap that these 2 could be merged, but if not, we should at least rename the methods to be specific.

File: components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt [hunk 4]
```diff
@@ -355,26 +348,87 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
...
+            val cpiCpkEntity: CpiCpkEntity = if(existingCpiCpkEntity != null) {
+                updateExistingCpiCpkEntity(existingCpiCpkEntity, cpk)
+            } else {
+                createNewCpiCpkEntity(cpi, cpk, cpkKey)
+            }
+            cpk.originalFileName!! to cpiCpkEntity
+        }
+    }
+
+    private fun createNewCpiCpkEntity(
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:369]: Is this needed? Can `CpiCpkEntity` not stay internalised to the entities? It makes the API a bit odd.

File: components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt [hunk 5]
```diff
@@ -355,26 +348,87 @@ class DatabaseChunkPersistence(private val entityManagerFactory: EntityManagerFa
...
+        cpk.metadata.fileChecksum.toString(),
+        CpkMetadataEntity(
+            cpkKey,
+            cpk.metadata.fileChecksum.toString(),
+            "1",
+            cpk.metadata.toJsonAvro()
+        )
+    )
+
+    private fun updateExistingCpiCpkEntity(
```
Comment (driessamyn) [inline components/chunking/chunk-db-write-impl/src/main/kotlin/net/corda/chunking/db/impl/persistence/DatabaseChunkPersistence.kt:392]: I don't think this belongs in this class as it's not specific about the upload.
